app/controllers/workflows_controller.rb

```diff
@@ -58,29 +56,33 @@ def stage # rubocop:todo Metrics/CyclomaticComplexity
     # else actually execute the task.
     unless params[:next_stage].nil?
       eager_loading = @task.included_for_do_task
-      @batch ||= Batch.includes(eager_loading).find(params[:batch_id])
+      @batch = Batch.includes(eager_loading).find(params[:batch_id])
 
-      unless @batch.editable?
-        flash[:error] = 'You cannot make changes to a completed batch.'
-        redirect_back fallback_location: root_path
+      editable, message = @task.can_process?(@batch, from_previous: true)
+
+      unless editable
+        redirect_back fallback_location: batch_path(@batch), alert: message
         return false
       end
 
       ActiveRecord::Base.transaction do
-        if @task.do_task(self, params)
+        task_success, task_message = @task.do_task(self, params)
+        if task_success
           # Task completed, start the batch is necessary and display the next one
           do_start_batch_task(@task, params)
           @stage += 1
           params[:id] = @stage
           @task = @workflow.tasks[@stage]
         end
+        flash[task_success ? :notice : :alert] ||= task_message if task_message
       end
     end
 
-    # Is this the last task in the workflow?
-    if @stage >= @workflow.tasks.size
+    if params[:commit] == 'Update'
```

Comment:
When editing batch steps you get taken back to the batch page, rather than on to the next step. This is mostly to avoid a user thinking they need to carry on processing the remainder of the batch.